flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -0,0 +1,296 @@
+
+	@Override
+	public void shutDown() throws Exception {
+		Exception exception = null;
+		// stop all currently running JobManagerRunners
+		for (JobManagerRunner jobManagerRunner : jobManagerRunners.values()) {
+			jobManagerRunner.shutdown();
+		}
+
+		jobManagerRunners.clear();
+
+		try {
+			submittedJobGraphStore.stop();
+		} catch (Exception e) {
+			exception = ExceptionUtils.firstOrSuppressed(e, exception);
+		}
+
+		try {
+			super.shutDown();
+		} catch (Exception e) {
+			exception = ExceptionUtils.firstOrSuppressed(e, exception);
+		}
+
+		if (exception != null) {
+			throw new FlinkException("Could not properly terminate the Dispatcher.", exception);
+		}
+	}
+
+	//------------------------------------------------------
+	// RPCs
+	//------------------------------------------------------
+
+	@RpcMethod
+	public Acknowledge submitJob(JobGraph jobGraph) throws JobSubmissionException {
+		final JobID jobId = jobGraph.getJobID();
+
+		log.info("Submitting job {} ({}).", jobGraph.getJobID(), jobGraph.getName());
+
+		final RunningJobsRegistry.JobSchedulingStatus jobSchedulingStatus;
+
+		try {
+			jobSchedulingStatus = runningJobsRegistry.getJobSchedulingStatus(jobId);
+		} catch (IOException e) {
+			log.warn("Cannot retrieve job status for {}.", jobId, e);
+			throw new JobSubmissionException(jobId, "Could not retrieve the job status.", e);
+		}
+
+		if (jobSchedulingStatus == RunningJobsRegistry.JobSchedulingStatus.PENDING) {
+			try {
+				submittedJobGraphStore.putJobGraph(new SubmittedJobGraph(jobGraph, null));
+			} catch (Exception e) {
+				log.warn("Cannot persist JobGraph.", e);
+				throw new JobSubmissionException(jobId, "Could not persist JobGraph.", e);
+			}
+
+			final JobManagerRunner jobManagerRunner;
+
+			try {
+				jobManagerRunner = createJobManagerRunner(
+					ResourceID.generate(),
+					jobGraph,
+					null,
+					getRpcService(),
+					highAvailabilityServices,
+					blobServer,
+					heartbeatServices,
+					metricRegistry,
+					new DispatcherOnCompleteActions(jobGraph.getJobID()),
+					fatalErrorHandler);
+
+				jobManagerRunner.start();
+			} catch (Exception e) {
+				try {
+					// We should only remove a job from the submitted job graph store
+					// if the initial submission failed. Never in case of a recovery
+					submittedJobGraphStore.removeJobGraph(jobId);
+				} catch (Throwable t) {
+					log.warn("Cannot remove job graph from submitted job graph store.", t);
+					e.addSuppressed(t);
+				}
+
+				throw new JobSubmissionException(jobId, "Could not start JobManager.", e);
+			}
+
+			jobManagerRunners.put(jobId, jobManagerRunner);
+
+			return Acknowledge.get();
+		} else {
+			throw new JobSubmissionException(jobId, "Job has already been submitted and " +
+				"is currently in state " + jobSchedulingStatus + '.');
+		}
+	}
+
+	@RpcMethod
+	Collection<JobID> listJobs() {
+		// TODO: return proper list of running jobs
+		return Collections.emptyList();
```

Comment:
could we not already implement this by taking the keyset from the jobManagerRunners map?